### PR TITLE
Add `EmbedReduce` layer.

### PR DESCRIPTION
**Pull request by hertschuh**
This layer looks up multiple embeddings per row and then combines the result to create a single embedding per row.

Also renamed `types.TensorShape` to `types.Tensor`.

### CODE REVIEW

File: keras_rs/src/layers/embedding/embed_reduce.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+
+        # Make weights and the unreduced embeddings have the same rank.
+        weights_rank = len(weights.shape)
+        if weights_rank < unreduced_rank:
+            weights = ops.expand_dims(
+                weights, axis=tuple(range(weights_rank, unreduced_rank))
+            )
+
+        # Note that `x` and `weights` are:
+        # - ragged if `inputs` was ragged and `weights` was ragged or None
```
Comment (abheesht17) [inline keras_rs/src/layers/embedding/embed_reduce.py:154]: How can a tensor be ragged since JAX and Torch do not have ragged support? Confused

**Review by abheesht17 — COMMENTED** [review]
Quick review

File: keras_rs/src/layers/embedding/embed_reduce.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+    """
+
+    def __init__(
+        self,
+        input_dim: int,
+        output_dim: int,
+        embeddings_initializer: types.InitializerLike = "uniform",
+        embeddings_regularizer: Optional[types.RegularizerLike] = None,
+        embeddings_constraint: Optional[types.ConstraintLike] = None,
+        mask_zero: bool = False,
```
[the remark this answers is not in the document]
Comment (hertschuh) [reply keras_rs/src/layers/embedding/embed_reduce.py:67]: Yes, to get:
- documentation
- code completion
- the same positional argument order as the super class.

File: keras_rs/src/layers/embedding/embed_reduce.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+    allowed. Weights can optionally be passed to the `__call__` method to
+    apply weights to different samples before reduction.
+
+    This layer supports sparse inputs and ragged inputs with backends that
+    support them. The output after reduction is dense. For ragged inputs, the
+    ragged dimension must be 1 as it is the dimension that is reduced.
+
+    Args:
+        input_dim: Integer. Size of the vocabulary,
+            i.e. maximum integer index + 1.
```
[the remark this answers is not in the document]
Comment (hertschuh) [reply keras_rs/src/layers/embedding/embed_reduce.py:33]: Not sure the highlight is showing correctly. Which sentence?

File: keras_rs/src/layers/embedding/embed_reduce.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+            If `mask_zero` is set to `True`, as a consequence,
+            index 0 cannot be used in the vocabulary (`input_dim` should
+            equal size of vocabulary + 1).
+        weights: Optional floating-point matrix of size
+            `(input_dim, output_dim)`. The initial embeddings values
+            to use.
+        combiner: Specifies how to reduce if there are multiple entries in a
+            single row. Currently `mean`, `sqrtn` and `sum` are supported.
+            `mean` is the default. `sqrtn` often achieves good accuracy, in
+            particular with bag-of-words columns.
```
[the remark this answers is not in the document]
Comment (hertschuh) [reply keras_rs/src/layers/embedding/embed_reduce.py:56]: Hmm... that wouldn't work because `ops.mean` doesn't know how to do a weighted mean, and it's not just about multiplying before doing the mean, the denominator is different. Also `"sqrtn"` doesn't exist as an op.

File: keras_rs/src/layers/embedding/embed_reduce.py [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+        self,
+        inputs: types.Tensor,
+        weights: Optional[types.Tensor] = None,
+    ) -> types.Tensor:
+        """Apply embedding and reduction.
+
+        Args:
+            inputs: 1D tensor to embed or 2D tensor to embed and reduce.
+            weights: Optional tensor of weights to apply before reduction, which
+               is 2D and the same shape as `inputs`.
```
[the remark this answers is not in the document]
Comment (hertschuh) [reply keras_rs/src/layers/embedding/embed_reduce.py:99]: Oh right, let me fix that. In fact it can also be 1D if the inputs are 2D.

Note that in theory you can even pass 3D weights with 2D inputs, meaning it multiplies each embedding value separately, which kind of doesn't make sense. But I won't document that.

File: keras_rs/src/layers/embedding/embed_reduce.py [hunk 6]
```diff
@@ -0,0 +1,188 @@
+from typing import Any, Optional
+
+import keras
+from keras import ops
+
+from keras_rs.src import types
+from keras_rs.src.api_export import keras_rs_export
+from keras_rs.src.utils.keras_utils import check_shapes_compatible
+
+SUPPORTED_COMBINERS = frozenset(["mean", "sum", "sqrtn"])
```
[the remark this answers is not in the document]
Comment (hertschuh) [reply keras_rs/src/layers/embedding/embed_reduce.py:10]: Yes, there are so few of them. That's feedback I got in a different context (to use `frozenset`).

File: keras_rs/src/layers/embedding/embed_reduce.py (hunk 1 above)
Comment (hertschuh) [reply keras_rs/src/layers/embedding/embed_reduce.py:154]: Oh, this if for TF only for now. This comment is specifically to explain that it's usually dense.

Note that this code was carefully crafted to handle dense, ragged and sparse without branching and without even knowing about the concept of ragged / sparse. (Part of it is we're missing APIs for that)

Comment (abheesht17) [reply keras_rs/src/layers/embedding/embed_reduce.py:155]: Ah, okay. I should have checked the unit tests

File: keras_rs/src/layers/embedding/embed_reduce.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (abheesht17) [reply keras_rs/src/layers/embedding/embed_reduce.py:33]: Looks like we can move line 33 up. Same in other places

image

[the remark this answers is not in the document]
Comment (hertschuh) [reply keras_rs/src/layers/embedding/embed_reduce.py:33]: Oh, the wrapping is off. I copied that from `Embedding`. Will fix.

File: keras_rs/src/layers/embedding/embed_reduce.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+        # With nothing to reduce, the "mean" and "sqrtn" combiners don't modify
+        # the result, but we still use the weights as a mask for the result for
+        # ragged and sparse.
+        # With the "sum" combiner, even without reduction, we still need to
+        # multiply by the weight to produce the correct result.
+        if weights is None or (unreduced_rank <= 2 and self.combiner != "sum"):
+            # Discard the weights if there were some and create a mask for
+            # ragged and sparse tensors to mask the result correctly (sparse
+            # only) and the apply the reduction correctly (ragged and sparse).
+            weights = ops.ones_like(inputs, dtype=dtype)
```
Comment (abheesht17) [inline keras_rs/src/layers/embedding/embed_reduce.py:132]: I didn't get this part. Why do we set the weights to all ones only for non-sum combiners?

**Review by abheesht17 — APPROVED** [review]
Thanks, left some NITs. Approved so as to not block dist. embedding

File: keras_rs/src/layers/embedding/embed_reduce_test.py [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+                    unique_indices=True,
+                )
+            else:
+                raise ValueError(
+                    f"Unsupported backend: {keras.backend.backend()}"
+                )
+        else:
+            raise ValueError(
+                f"Unsupported type/rank combination: {input_type} {input_rank}d"
+            )
```
[the remark this answers is not in the document]
Comment (hertschuh) [reply keras_rs/src/layers/embedding/embed_reduce_test.py:88]: Oh this was cider-v complaining that this case was not handled, but I can remove it here.

File: keras_rs/src/layers/embedding/embed_reduce_test.py [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+                )
+                weights = jax_sparse.BCOO(
+                    ([1.0, 1.0, 2.0, 3.0, 4.0], indices),
+                    shape=(2, 4),
+                    unique_indices=True,
+                )
+            else:
+                raise ValueError(
+                    f"Unsupported backend: {keras.backend.backend()}"
+                )
```
[the remark this answers is not in the document]
Comment (hertschuh) [reply keras_rs/src/layers/embedding/embed_reduce_test.py:84]: Oh this was cider-v complaining that this case was not handled, but I can remove it here.

File: keras_rs/src/layers/embedding/embed_reduce_test.py [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+from absl.testing import parameterized
+from keras import ops
+from keras.layers import deserialize
+from keras.layers import serialize
+
+from keras_rs.src import testing
+from keras_rs.src.layers.embedding.embed_reduce import EmbedReduce
+
+
+class EmbedReduceTest(testing.TestCase, parameterized.TestCase):
```
[the remark this answers is not in the document]
Comment (hertschuh) [reply keras_rs/src/layers/embedding/embed_reduce_test.py:13]: Will do in a separate PR.

File: keras_rs/src/layers/embedding/embed_reduce.py (hunk 7 above)
Comment (hertschuh) [reply keras_rs/src/layers/embedding/embed_reduce.py:132]: 1. When `weights` is `None`:
    - For ragged inputs, after embedding, we get a ragged result that has a ragged dimension of 1, but when we do the mean or sqrtn, we need to divide by the number of items in each row. Because there is no explicit cross backend API to get the row length. `ones_like` gives us a ragged tensor that is ragged in the same way as the inputs. When we do `ops.sum(weights, axis=-2)`, it gives us the number of items per row.
    - For sparse inputs, after embedding, we get a dense tensor, not a sparse tensor. What it does for missing values is use embedding 0. These are bogus embedding and should be ignored. `ones_like` gives us a sparse tensor with the exact same missing values. Later, when we do `x = ops.multiply(x, weights)`, which masks the bogus values (note that `weights` has been densified beforehand). Additionally, when we do `ops.sum(weights, axis=-2)`, it gives us the number of items per row.

2. When `unreduced_rank <= 2`, this means that the inputs where 1D, there is only one embedding per row, so there is no real reduction going on.
    - For mean: `result = weights * x / weights = x` we don't needs the weights
    - For sqrtn: `result = weights * x / sqrt(square(weights)) = x` we don't needs the weights
    - For sum however: `result = weights * x` so we do need the weights
 So for mean and sqrtn we don't need the weights, it's not an optimization, it's more of a way to avoid divisions by zero and improve precision. In fact unit tests fail without it because of the imprecisions introduced.

I'll add some of that in the comments.

Comment (abheesht17) [reply keras_rs/src/layers/embedding/embed_reduce.py:151]: Thanks!
